src/main/java/net/openhft/chronicle/core/threads/ThreadLock.java

```diff
@@ -0,0 +1,134 @@
+package net.openhft.chronicle.core.threads;
+
+import net.openhft.chronicle.core.Jvm;
+import net.openhft.chronicle.core.OS;
+import net.openhft.chronicle.core.values.LongValue;
+
+import java.io.File;
+
+public class ThreadLock {
+    static final Metrics METRICS = new Metrics();
+    private static final long TOP_BITS = ~0L << 32;
+    private static final int UNLOCKED = 0;
+    private final LongValue twoThreadId;
+    private final long timeoutMs;
+    int busyLoopCount = 20_000;
+    int busyLockSlowerCount = 100;
+
+    public ThreadLock(LongValue twoThreadId, long timeoutMs) {
+        this.twoThreadId = twoThreadId;
+        this.timeoutMs = timeoutMs;
+    }
+
+    public boolean tryLock(int threadId) {
+        if (threadId == 0)
+            throw new IllegalArgumentException();
+        long twoThreadId0 = twoThreadId.getVolatileValue();
+        int threadId0 = (int) twoThreadId0;
+        if (threadId0 != UNLOCKED) {
+            if (threadId0 == threadId)
+                throw new IllegalStateException("trying to lock twice");
+            return false;
+        }
+        return twoThreadId.compareAndSwapValue(twoThreadId0, (twoThreadId0 & TOP_BITS) | threadId);
+    }
+
+    public void lock(int threadId) throws InterruptedRuntimeException {
+        if (tryLock(threadId))
+            return;
+        busyLock(threadId);
+    }
+
+    private void busyLock(int threadId) {
+        // busy wait
+        for (int i = 0; i < busyLoopCount; i++) {
+            if (tryLock(threadId))
+                return;
+            Jvm.nanoPause();
+        }
+        if (busyLockSlower(threadId))
+            return;
+        forceUnlockAndRetry(threadId);
+    }
+
+    @SuppressWarnings("java:S3776")
+    private boolean busyLockSlower(int threadId) throws InterruptedRuntimeException {
+        long endMS = System.currentTimeMillis() + timeoutMs;
+        do {
+            for (int i = 0; i < busyLockSlowerCount; i++) {
+                if (tryLock(threadId))
+                    return true;
+                Thread.yield();
+            }
+            if (METRICS.supportsProc) {
+                int lockedThread = (int) twoThreadId.getVolatileValue();
+                if (lockedThread == 0)
+                    continue;
+                if (!isThreadRunning(lockedThread)) {
```

Comment:
There appears to be a race condition here. 
If thread A holds the lock and is dead, and threads B and C try to acquire the lock, both B and C detect A is dead here.
B and C then both go to forceUnlockAndRetry. But forceUnlockAndRetry unconditionally clears the current owner.
So assuming B gets there first, B (correctly) clears A and acquires the lock
C then (incorrectly) clears B immediately after

To prevent this, busyLockSlower should return the "assumed dead" thread id to the caller. This is then explicitly passed to forceUnlockAndRetry - so forceUnlockAndRetry is told on entry which (dead) thread is believed to be holding the lock and needs to be killed, rather than reading the current value (which may now be a non-dead thread) at line 85.

Reply:
Similarly on the timeout case: A is alive but times out, B and C both proceed to force unlock A. B unlocks A, but C then immediately unlocks B.

Reply:
(Strictly speaking, forceUnlockAndRetry should return the full twoThreadId to be force unlocked, so it can be used directly in the cas).

Reply:
Going further, rather than forceUnlock followed by a separate lock attempt, when passing in the known explicit dead/timedout twoThreadId both steps can be combined in a single atomic CAS: force-unlock-and-acquire by CASing on the old twoThreadId and setting the new value as (twoThreadID<<32 | newThreadId)

Reply:
I suggest creating an interface for the intended functionality and then providing a static constructor method that will return an implementation located in the internal package.

Reply:
I have changed forceLock to use a retained "old" lock value which timed out. If the value has changed in the meantime the force lock fails and it tries again.